core/src/main/scala/spire/math/Algebraic.scala

```diff
@@ -1,64 +1,165 @@
 
-import spire.algebra.{Eq, EuclideanRing, Field, IsReal, NRoot, Order, Ring, Sign, Signed}
-import spire.algebra.Sign.{ Positive, Negative, Zero }
-import java.math.{ MathContext, BigInteger, BigDecimal => BigDec }
-import scala.math.{ ScalaNumber, ScalaNumericConversions }
+  /**
+   * Returns an `Int` with the same sign as this algebraic number. Algebraic
+   * numbers support exact sign tests, so this is guaranteed to be accurate.
+   */
+  def signum: Int = expr.signum
 
-import algebraic._
+  /**
+   * Returns the sign of this Algebraic number. Algebraic numbers support exact
+   * sign tests, so this is guaranteed to be accurate.
+   */
+  def sign: Sign = Sign(signum)
 
+  /**
+   * Return a non-negative `Algebraic` with the same magnitude as this one.
+   */
+  def abs: Algebraic =
+    if (this.signum < 0) -this else this
 
-/**
- * An general Algebraic type. Can be used represent real numbers and approximate
- * them on-demand.
- */
-@SerialVersionUID(0L)
-final class Algebraic private (val expr: Expr[Algebraic])
-extends ScalaNumber with ScalaNumericConversions
-   with RealLike[Algebraic]
-   with BMFSSBound[Algebraic]
-   with BigDecimalApprox[Algebraic]
-   with FPFilter[Algebraic]
-   with ConstantFolder[Algebraic]
-   with BubbleUpDivs[Algebraic]
-   with PrettyToString[Algebraic]
-   with Ordered[Algebraic]
-   with Serializable {
-
-  val coexpr: Coexpr[Algebraic] = Algebraic.RealCoexpr
-
-  // ugh
-  override def byteValue(): Byte = fpf.toLong map (_.toByte) getOrElse super.toByte
-  override def shortValue(): Short = fpf.toLong map (_.toShort) getOrElse super.toShort
+  def unary_- : Algebraic =
+    new Algebraic(Expr.Neg(expr))
+
+  def +(that: Algebraic): Algebraic =
+    new Algebraic(Expr.Add(this.expr, that.expr))
+
+  def -(that: Algebraic): Algebraic =
+    new Algebraic(Expr.Sub(this.expr, that.expr))
+
+  def *(that: Algebraic): Algebraic =
+    new Algebraic(Expr.Mul(this.expr, that.expr))
+
+  def /(that: Algebraic): Algebraic =
+    new Algebraic(Expr.Div(this.expr, that.expr))
+
+  /**
+   * Returns an `Algebraic` whose value is just the integer part of
+   * `this / that`. This operation is exact.
+   */
+  def quot(that: Algebraic): Algebraic =
+    this /~ that
+
+  /** An alias for [[quot]]. */
+  def /~(that: Algebraic): Algebraic =
+    Algebraic((this / that).toBigInt)
+
+  /**
+   * Returns an `Algebraic` whose value is the difference between `this` and
+   * `(this /~ that) * that` -- the modulus.
+   */
+  def mod(that: Algebraic): Algebraic =
+    this % that
+
+  /** An alias for [[mod]]. */
+  def %(that: Algebraic): Algebraic =
+    this - (this /~ that) * that
+
+  /** Returns the square root of this number. */
+  def sqrt: Algebraic = nroot(2)
+
+  /** Returns the cube root of this number. */
+  def cbrt: Algebraic = nroot(3)
+
+  /** Returns the `k`-th root of this number. */
+  def nroot(k: Int): Algebraic = if (k < 0) {
```

Comment:
We should probably test this with very large values (namely > 306) and make sure it's OK. I'm not sure how slow this will be. I think just testing this with constants (or very simple expressions) should be fine for now, since we're not testing how the underlying errors propagate.